compiler/rustc_interface/src/util.rs

```diff
@@ -528,6 +528,18 @@ pub(crate) fn check_attr_crate_type(
                         );
                     }
                 }
+            } else {
+                // This is here mainly to check for using a macro, such as
+                // #![crate_type = foo!()]. That is not supported since the
+                // crate type needs to be known very early in compilation long
+                // before expansion. Otherwise, validation would normally be
```

Comment:
>crate type needs to be known very early in compilation long before expansion

Why tho?
This information seems irrelevant to parser or macro expander.

Reply:
Ah, I see `StableCrateId::new` uses the crate types, although the use seems incorrect.

Otherwise it's only used before proc macro harness generation (to determine whether generate it or not), which happens only after all macros are expanded.

In other words, this is something that needs to be investigated.

Reply:
cc @bjorn3 
This is related to https://github.com/rust-lang/rust/pull/83744 (for which I don't see much motivation).

Reply:
> Ah, I see StableCrateId::new uses the crate types, although the use seems incorrect.

This is necessary for backwards compatibility. Without it the following would give an error about a duplicate `StableCrateId`:

```bash
$ echo "" | rustc - --crate-type lib --crate-name foo
$ echo "extern crate foo; fn main() {}" | rustc - --crate-type bin --crate-name foo -L .
```

the error would be (forced it by using lib as crate type in both cases):

```
error[E0519]: the current crate is indistinguishable from one of its dependencies: it has the same crate-name `foo` and was compiled with the same `-C metadata` arguments. This will result in symbol conflicts between the two.
 --> <anon>:1:1
  |
1 | extern crate foo; fn main() {}
  | ^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

I would personally be happy if it were to be turned into an error as it complicates things, but it is a breaking change.

> Otherwise it's only used before proc macro harness generation (to determine whether generate it or not), which happens only after all macros are expanded.

Currently expansion has to happen before the `StableCrateId` can be set in the `Session` due to the dependency on expansion to determine all `#![crate_type = "..."]`. Once the future compat lint in #83744 becomes a hard error, it becomes possible to determine the final crate name and crate type for certain without any expansion by only parsing the crate level attributes, thus allowing the `StableCrateId` to be set immediately while creating the `Session`.

Reply:
>allowing the `StableCrateId` to be set immediately while creating the `Session`

Are there specific reasons to want `StableCrateId` be created together with the `Session`?

Reply:
@bjorn3 
Could you copy this motivation into the description of https://github.com/rust-lang/rust/pull/83744?

Reply:
Done